### PR TITLE
Fix #1508 - the `needs()` function should propagate calling task's params/args down to needed tasks

### LINKED ISSUE #1508: The `needs()` function does not propogate run-time task arguments (`%params` and `@args`) from the calling task down to the "needed" tasks
#### Edit: 

As it turns out: this bug is really easy to fix, as it appears to stem from a typo. I am preparing a PR with accompanying tests.

## Describe the bug

The `needs()` function has a bug which prevents it from implicitely propogating the calling task's params/args down to the "needed" tasks when it _runs_ them.

## How to reproduce it

1. In a succint `Rexfile`, define two simple tasks (`task_a` & `task_b`)  that dump their run-time arguments (`@_`) on `STDERR`
2. Make sure that `task_b` calls `needs("task_a")`
3. From the shell, launch `rex task_b` (with appropriate host & auth info) and observe the results
    The dump from `task_a` is empty (consisting of empty containers `[{},[]]`), whereas `task_b` normally displays its `params`.  
4. Also try `rex task_b --greetings=Hello`, and observe that the result is similar to the previous trial.

#### Rexfile
```perl
# Rexfile
### ex: set ft=perl :

use Rex -feature => [qw(1.13 exec_autodie)];
use DDP; # alias for `Data::Printer`, used just for pretty display of dumps.

desc "Display host name";
task hostname => sub {
  say STDERR "hostname: \@_ : "; DDP::p(@_);
  say STDERR ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>";   
  say ''. run('hostname');
};

desc "Display host info";
task hostinfo => sub {
  say STDERR "hostinfo: \@_ : "; DDP::p(@_);
  say STDERR ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>";   
  needs "hostname";
};

1;

```

#### Run the example from the shell 

Execute `rex` within the directory where the above `Rexfile` resides (with appropriate host and authentication switches).

```text
$ rex hostinfo --greetings=Hello
[2021-09-26 18:26:45] INFO - Running task hostinfo on medusa
hostinfo: @_ : 
[
    [0] {
            greetings => "Hello",
            hostinfo  => 1
        },
    [1] []
]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
hostname: @_ : 
[]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
medusa
```

## Expected behavior

The `needs()` function should implicitly propagate the calling task's params/args down to the "needed" tasks when it _runs_ them, as documented and seemingly intended by the current code base (see notes below).

## Circumstances

 - Rex version: (R)?ex 1.13.4 (also on latest branch master)
 - Perl version: v5.34.0
 - OS running rex: ~~~ (irrelevant)
 - OS managed by rex: ~~~ (irrelevant)
 - How rex was installed: git clone repo 

## Notes

It turns out: this is really easy to fix, as it appears to stem from a typo.

### The culprit

Here's the offending portion of code in `Rex::Commands`.

```perl

# lib/Rex/Commands.pm 
...
sub needs {
    ...
    if ( @args && grep ( /^\Q$task_name\E$/, @args ) ) {
      Rex::Logger::debug( "Calling " . $task_o->name );
      $task_o->run( "<func>", params => \@task_args, args => \%task_opts );
    }
    elsif ( !@args ) {
      Rex::Logger::debug( "Calling " . $task_o->name );
      $task_o->run( "<func>", params => \@task_args, args => \%task_opts );
    }
    ...
}

```

Notice how `params` and `args` are mixed up (**interchanged**) when being passed to `run`!! This certainly looks like a typo.

The current test suite does not catch the issue. The tests in `needs.t` do not cover propagation of parameters/args.

### Quick fix

For a fix with minimal changes, just patch the **two occurrences** of the `$task_o->run` call, within the body of the `needs()` subroutine (in `Rex::Commands`), as below:

    - $task_o->run( "<func>", params => \@task_args, args => \%task_opts );
    + $task_o->run( "<func>", params => \%task_opts, args => \@task_args );

### Alternative fix (slightly better, imho)

There does not appear to be any particular reason for having  **two identical invocations** of the `->run` method in the above code snippet.

So, the offending code can be replaced by the following -logically equivalent- snippet. I will put that in a separate commit; in case there are other (historical?) reasons to keep the two branches of the if statement around.

In any case, all tests pass either way.

```perl

# lib/Rex/Commands.pm

...

sub needs {
    ...
    if ( !@args or @args && grep ( /^\Q$task_name\E$/, @args ) ) {
      Rex::Logger::debug( "Calling " . $task_o->name );
      $task_o->run( "<func>", params => \%task_opts, args => \@task_args );
    }
    ...
}

```

### PR DESCRIPTION
**Pull request by tabulon**
# PR for fixing #1508 - the `needs()`function should propagate calling task's params/args down to needed tasks

Hi,

This PR is an attempt to fix #1508 by correcting what appears to be a typo (or copy/paste error) in the body of the `needs()` function in `Rex::Commands`.

The bug appears to have been introduced initially by PR #1157 (the fix for #1066 ) with 48c737b.

Notice how params and args were mixed up (interchanged) when being passed down  -- within the `needs()` function.

Two alternate fixes are proposed (in commits 2 & 3), which may be cherry-picked, or just applied in series, which amounts to picking the latter.

Please review and merge, or let me know how to improve it further.

## Open questions

While this is clearly bug fix, it can't be ruled out that some code out in the wild may have come to rely on the incorrect behavior during the time of its existence.
(since around 2017).

Therefore, it might perhaps be safer to introduce a "feature switch" for introducing this fix... if the maintainers deem that necessary.

This PR does NOT include such a feature switch, as I guess (perhaps wrongly) that most people might have already turned to using `run_task` in the mean time, 
because of glitches like this with the `need()` function, some of which appear to have already been handled by #1188, but not this one.

That's only a guess, though...

## Bundled commits

1. Add tests (marked as TODO) related to the bug (#1508). See below 
2. Fix #1508, by correcting a typo in a couple places       _(quick fix, with minimal changes in code)_
3. Fix #1508 by correcting a typo + minimal code cleanup    _(alternate fix)_
4. Update ChangeLog, remove "TODO" mark from tests related to #1508

## How to test

```bash
$ prove -v t/issue/1508.t   # for this issue
$ prove t/**/*.t            # for non-regression
```

As long as the related tests remain marked as "TODO", 
they will not report failures during normal test runs.

To see their true pass/fail status, you have to pass
the '-v' option to `prove`.

The last commit removes the "TODO" mark from tests. 
Once that commit is merged, the '-v' switch is no longer needed.

## Checklist

- [x] based on top of latest source code 
- [x] changelog entry included 
- [ ] tests pass in CI 
- [x] git history is clean 
- [x] git commit messages are [well-written](https://chris.beams.io/posts/git-commit/#seven-rules)

### CODE REVIEW

**Comment by ferki** [comment]
Thanks for this PR, @tabulon!

I'm going to comment on your points first below before doing a full review.

---

> While this is clearly bug fix, it can't be ruled out that some code out in the wild may have come to rely on the incorrect behavior during the time of its existence. (since around 2017).

Also, code in the wild might have relied on the original behavior for years before 2017 :) The original author seems to have intended options and arguments to be passed around by `needs`. So I agree this is a bugfix to restore an unintentionally changed behavior for backwards compatibility reasons.

> This PR does NOT include such a feature switch, as I guess (perhaps wrongly) that most people might have already turned to using `run_task` in the mean time, because of glitches like this with the `need()` function, some of which appear to have already been handled by #1188, but not this one.

I don't think a new feature flag is necessary right now. If we have to make the "broken behavior" available again, that could be an opt-in, though.

I also believe that a lot of users of `needs` do not rely on passing around options and arguments between tasks (but their tasks are getting their input from elsewhere, e.g. CMDB). I also always solely relied on the "call this other task on the current connection" aspect.

> **Note**: Sorry about the last commit... which is only there to satisfy `xt/author/perltidy.t`... Please don't make me redo the entire PR just because the tidying comes in a separate commit

I don't think the PR needs to be re-done :) I might look into an opportunity to squash away the tidy-only commit at merge-time.

> The last commit removes the "TODO" mark from tests.

If I understand you correctly, the tests were treated as TODO first because you weren't sure in advance which of the fixes would be cherry-picked. In the spirit of TDD's "red-green-refactor" it's fine to do it something like this next time:

- write a failing test first and push it (as draft PR)
- push the minimum amount of change needed to fix (so it passes now)
- refactor the code to simplify it (=remove the duplication)

File: t/issue/1508.t [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  package T; # Helper package (for cutting down boilerplate in tests)
+  use Storable;
+
+  sub track_taskrun {
+    my %opts = ref $_[-1] eq 'HASH'
+      ? %{
+      ;
+      pop
+      }
+      : ();
```
Comment (ferki) [inline t/issue/1508.t:33]: Based on the calls in this file, it seems that the first argument is always a scalar value, and the last argument is always a hash reference. Does it make sense to still keep this unpack logic around?

Maybe this works too:

```perl
my ($called_task, %opts) = @_;
```

That way we don't need the for loop either. What do you think?

(Same goes for `check_needed` below)

File: t/issue/1508.t [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+use Test::More;
+use Rex::Commands;
+
+{
+
+  package T; # Helper package (for cutting down boilerplate in tests)
+  use Storable;
+
+  sub track_taskrun {
```
Comment (ferki) [inline t/issue/1508.t:27]: If I understand correctly this subroutine saves the arguments passed to it in a file via Storable. If that's true, would it be better to call it something like `save_arguments`?

File: t/issue/1508.t [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    my $argv      = $opts{argv};
+    my $do_unlink = delete $opts{unlink} // 1;
+    my @prereqs   = (@_);
+
+    for (@prereqs) {
+      my $file = "${_}.txt";
+
+      -f "$file" or die;
+      my $propagated_argv = Storable::retrieve("$file");
+      Test::More::_deep_check( $propagated_argv, $argv ) or die;
```
Comment (ferki) [inline t/issue/1508.t:58]: What are the advantages of using the private `_deep_check` explicitly here instead of the public `is_deeply`?

**Review by ferki — CHANGES_REQUESTED** [review]
I believe the code part is correct, but we probably need to find a better way to test the behavior for passing options and arguments around. I think we could keep working collaboratively on this PR and squash away the new changes into the original ones and/or add follow-up commits.

I also think I've found a bug in the author tests which caused the test file to be not checked by the Perl::Critic tests (e.g. `Modules::ProhibitMultiplePackages` should be in effect but this has passed with multiple packages in it). I'm going to fix that separately.

**Comment by ferki** [comment]
Alright, the perlcritic tests should be solved by now via #1516.

Please [rebase](https://www.w3docs.com/snippets/git/how-to-rebase-git-branch.html) these changes on top of the default branch in order to test them against the current state.

After the rebase, it should show some failures about the test file, which should also give us some ideas about how to go about testing.

File: t/issue/1508.t [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+run-time task arguments (%params and @args) from the calling task down to the "needed" tasks.
+
+=head1 DETAILS
+
+  * AUTHOR / DATE : [tabulon]@[2021-09-26]
+  * RELATES-TO    : [github issue #1508](https://github.com/RexOps/Rex/issues/1508#issue-1007457392)
+  * INSPIRED from : t/needs.t
+
+=cut
+
```
[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:18]: Hi @Ferki, 

Thank you for your through review -- of this PR as well as the related issue. 

I will try to briefly address your remarks in several chunks.

### POD in test files

Since the `.t` files are valid perl scripts, putting `POD` is technically allowed. 

For some kinds of content, I find it nicer than block/line comments. It renders better and folds neatly in the text editor. Whether or not it is socially allowed is up to the maintainer (you), so go ahead and yank it if you wish.

File: t/issue/1508.t [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
```
[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:1]: As mentioned, and despite how it may look at first, this test script was heavily inspired from `t/needs.t`. 

It's basically a refactored version of it, with one notable functional difference: unlike  `t/needs.t`, this one also checks the run-time task options and params. 

So, my initial -joking- response would be : 

sure, please walk us through `t/needs.t` and I will do the same for this one :-)

[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:1]: Joking aside, here's a brief run-through :

Just like in `t/needs.t`, the `main` package does the following : 

- first, define a bunch of tasks : 

```perl
task test => sub {...}
task test2 => sub {...}
...
task test6 => sub {...}
```
- Then, run each task and check that it has returned successfully (which constitutes the actual "test")

Just like in `t/needs.t`, some of those tasks invoke the `needs()` function, either with other tasks in the `main` package, or  tasks defined in other packages (`Rex::Module` or `Nested::Module`) which also reside in the same test script.

#### 1) Why do we have  the two other packages (`Rex::Module` and `Nested::Module`) ? 

Same reason as we have them in `t/needs.t`, i.e. we want to check that the `needs()` function works across package namespaces.

[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:1]: #### 2) Why do we have an extra helper package (T), unlike in `t/needs.t` ?

Well, that's where the DRY refactoring happens... 

Notice how almost the same code is repeated again and again in the task subs in `t/needs.t`. 

Here, that boilerplate is replaced with calls to `track_taskrun` and `check_needed` which sensibly do the same, i.e. : 

- `track_taskrun`  :  When a task gets run, mark it as such by writing to a file with that name (and also save the its arguments there).

- `check_needed` : 
  * Make sure a "needed" task has already been run, i.e. : 
    * `die` unless a file with that name exists
    * `die` unless the arguments that were saved in that file correspond to the expected arguments
  * Also do some clean up, by deleting the said file. 
    Doing this here might not be considered very orthodox, but that's also what's happening in `t/needs.t`)

Notice how failure is signaled with `die`,  just like in `t/needs.t`. 

This also partly explains why we employ the private `_deep_check()`  routine from `Test::More::` here.

Just like you, my first reflex was employing `is_deeply`, but that doesn't work because "Test::More" chokes as it is then unable to keep track of the actual number of tests. 

The cleaner alternative would have been to rely on  `is_deeply` in the `main` test loop `for my $task ( $run_list->tasks )` in the test script's package `main`, but that won't work either, because the summary info (from `$task_list->get_summary`) does not contain task params/arguments... Well it probably should... :-)

[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:1]: #### 3) Why do we have to do file IO in the first place ?

I was also suprised by this at first. I guess the author of  `t/needs.t` could have the best answer here. 

The explanation I came up with has to do with the task execution model in Rex (each task possibly having separate SSH connection, possibly running on a separate thread/process, ...).

A no-brainer method for eliminating the hairiness that would get involved is to go through file IO.

`t/needs.t` does this quite simply by writing/truncating a file that has the same name of the needed task when its being run, and then merely checking the existence of the file from the calling task.

Here, we also keep track of the params/arguments, so we use `Storable` to do that, which also eliviates the need for manually opening and reading/writing to files.

I chose `Storable` because of its simplicity and popularity. Besides, `Rex` already has a runtime dependency on `Storable`, so using it in a test did not entail adding a dependency.

[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:1]: 3) Why the unpacking logic ?

At one point you ask if we could get rid of the unpacking logic in helper routines, which was actually neatly placed on a single line (before perl tidy messed it up :-) :

```perl
    my %opts = ref $_[-1] eq 'HASH' ? %{; pop } : ();
```
Yes, we could; by directly assuming a HASH reference as the last argument, like such : 

```perl
    my %opts = %{; pop };
```
But, if you know of a way of preventing perl-tidy doing its thing, I think it would be preferable to keep the unpacking logic in there. 

It gives the caller more freedom. That way, if desired, the current tests in `t/needs.t` can easily be implemented with those two refactored routines (`track_taskrun` and `check_needed`). 

The above also gives away my thinking about you suggestion of merging these tests into `t/needs.t`. 

Imho, the easiest and most maintainable way of doing that would be retrofit `t/needs.t` tests into this -refactored- model.

Otherwise, if we try to it the non-DRY way `t/needs.t` does this (by copying and pasting the file IO code in all tasks), we could easily end up with a mess of several hundred lines (if not more) of hard-to-maintain code...

File: t/issue/1508.t (hunk 2 above)
Comment (tabulon) [reply t/issue/1508.t:27]: If you wish so...

It also saved the arguments, yes. 
But its principal function is to register the fact that a given task has just been executed (also saving the run-time arguments, in our case).

File: t/issue/1508.t (hunk 1 above)
Comment (tabulon) [reply t/issue/1508.t:33]: Perhaps... 

The intention was the ability to do the following, when you have more than one needed task : 

```perl
needs ("task1", "task2", "task3")
check_needed("task1", "task2", "task3", \%opts) 
````
Currently, it looks like we don't have that case (with several needed tasks) in the tests. Maybe we should ?

Comment (tabulon) [reply t/issue/1508.t:33]: For `track_taskrun`, it's an easier call; yes we could get rid of some of the logic, as you suggest.

File: t/issue/1508.t (hunk 3 above)
Comment (tabulon) [reply t/issue/1508.t:58]: Replied elsewhere

File: t/issue/1508.t (hunk 4 above)
[the remark this answers is not in the document]
Comment (tabulon) [reply t/issue/1508.t:18]: ##### Why were the tests marked as TODO ?

The cherry-picking thing might have played a role in there, yes. 
But that's also what I usually try to do, because I find it way cleaner way doing TDD. 

If `Rex` policy is as you describe, I will try to remember next time. It's true that this has become quite popular among many languages lately.

Meanwhile, here are my reasons, in a nutshell, in an attempt to convince you of the added value:

* Marking tests that are expected to fail (as such) is an extra piece of valuable information.

* Perl (with its excellent Test::* modules) had the ability to do TDD way before other languages. TODO tests were invented as part of this scheme, and they have quite interesting characteristics. 

* When you do a normal `prove`, TODO tests still get executed (and their results are briefly reported), but their failure won't cause the whole test suite to fail!

* When you do a `prove -v` (verbose mode), TODO tests behave as as normal tests : they will get detailed reporting 

* If TODO tests actually pass (instead of failing), that's also reported as a surprise (so that you can suspect something fishy)

* Also, the cherry-picking of solutions can then happen across PRs. 

Basically, it tends to be a superior form of TDD, in my humble opinion: 

- You can start with writing TODO tests for a given issue (without necessarily committing any code that satisfies them) ; keep them around, even ship them as such. 
- Months later, you (or someone else) can come up with one or more solutions which can easily be checked with `prove -v`. 
- Once satisfied with a solution, you unmark the TODO tests within the PR that commits the solution.

And, of course, all of that can happen within the same PR (as was the case here).

But you probably already know all this... And you probably have good reasons to prefer otherwise.

In the end, whatever the maintainer of the project decides, of course.

**Comment by tabulon** [comment]
> Please [rebase](https://www.w3docs.com/snippets/git/how-to-rebase-git-branch.html) these changes on top of the default branch in order to test them against the current state.
> 
> After the rebase, it should show some failures about the test file, which should also give us some ideas about how to go about testing.

I will try to `rebase` as you suggest as soon as I get a chance.
Meanwhile, please note that `t/needs.t` also has multiple packages in it (all except the helper package `T`).

File: t/issue/1508.t (hunk 4 above)
[the remark this answers is not in the document]
Comment (ferki) [reply t/issue/1508.t:18]: # POD in tests

In general I'm not opposed to add POD to test files, provided we can realistically commit to establish it as the new norm, _and_ also to start adding them retroactively for consistency. I feel that would involve a considerable overhead, so I'd wish to discuss that topic outside of the context of this PR (and focus on passing arguments to `needs` here).

---

# TODO tests

Having new tests that first pass, which then start to fail when I successfully implement their logic feels backward _for my brain_ :upside_down_face: I can accept it works better for you, and I appreciate your detailed reasoning :heart: 

So far I used `TODO` for tests that are expected to fail, _and_ not intend to fix in the same PR that introduces them. So I'm fine shipping TODO tests (e.g. there's one in `t/file.t`), and I'd even prefer it over things like "skip this on Windows" whenever possible (the current test suite mostly SKIPs though :/ ).

In the usual PR context, I like to have it demonstrated by the CI run that the new tests are actually failing before the related change gets implemented. So I push the "Add tests for X" commits early, and watch them fail ("red"). Then I also like to demonstrate that the new commit (or sometimes series of commits) I push to implement/fix the failing behavior actually makes the test suite to pass ("green"). Then if I need to simplify the code, I can make that at the end, and demonstrate the tests still pass ("refactor")

If my PR would be about adding new tests that I don't intend to fix in the same PR, I think I'd still push a commit with the failing tests first, then make the CI pass with a second commit that marks them as TODO.

I'd be happy to further discuss these methodology details via [GitHub discussions](https://github.com/RexOps/Rex/discussions) or chat.

File: t/issue/1508.t (hunk 5 above)
[the remark this answers is not in the document]
Comment (ferki) [reply t/issue/1508.t:1]: > sure, please walk us through t/needs.t and I will do the same for this one :-)

Yeah, later I realized it is based on `t/needs.t`. My take is that `t/needs.t` is bad too, but that's part of the historical ballast from 10 years ago. That's a considerable amount of legacy to clean up, so that's why we use Test::Perl::Critic::Progressive to make sure we at least don't add more violations on top of the current pile :) These tests were unknowingly broken for PRs coming from forks, which was my mistake and I fixed that now (so it should properly complain about `t/issue/1508.t` after a rebase).

In other words, the policy is "we accept our legacy, but we don't want to add more ballast; therefore new code should be clean, or clean up old mess".

[the remark this answers is not in the document]
Comment (ferki) [reply t/issue/1508.t:1]: Thanks for the detailed explanation!

---

>  1) Why do we have the two other packages (Rex::Module and Nested::Module) ?

Rex chops off the leading `Rex::` part for task names for brevity. We have these two modules defined in `t/needs.t` to make sure `needs` work with namespaces both with and without the `Rex::` prefix.

---

> 2) Why do we have an extra helper package (T), unlike in t/needs.t ?

> This also partly explains why we employ the private _deep_check() routine from Test::More:: here.
>
> Just like you, my first reflex was employing is_deeply, but that doesn't work because "Test::More" chokes as it is then unable to keep track of the actual number of tests.

That sounds like a good indicator to invent a cleaner approach than the 10-year-old legacy `t/needs.t` - or even better, fix `t/needs.t` first :)

I wonder if we could split `t/needs.t` up into real modules under e.g `t/lib` first as a pure refactor, then modify it to support running the same test tasks both with and without arguments :thinking: 

---

> 3) Why do we have to do file IO in the first place ?

Yup, I see we use files to keep track of some execution data, and I don't have a much better way for that currently. `Storable` also sounds good for dumping/restoring the extra data needed by the argument tests. It is also a Perl core module, so it wouldn't be a new dependency anyway :+1: 

---

> 3. Why the unpacking logic ?

> was actually neatly placed on a single line (before perl tidy messed it up :-)

That might also be an indicator of weirdness :) I'm happy to tune perltidy rules and reformat the codebase with that.

In this case though perlcritic complains about the null statement (and in the exploded formatting, about the missing final semicolon in the block). Perhaps it would be more readable as fully unpacking @_ first, and then do this transformation logic on the last element of the passed array?

File: t/issue/1508.t (hunk 1 above)
Comment (ferki) [reply t/issue/1508.t:33]: > Currently, it looks like we don't have that case (with several needed tasks) in the tests. Maybe we should ?

I think that's what confused me initially. We don't have the multiple-tasks case tested now, so it might be premature optimizing.

But your're right, in the long run we probably should cover that use case too :+1:

File: t/issue/1508.t (hunk 2 above)
Comment (ferki) [reply t/issue/1508.t:27]: I think by now I better understand the original name. This is a subroutine that tracks task execution (~run).

**Comment by ferki** [comment]
> Meanwhile, please note that t/needs.t also has multiple packages in it (all except the helper package T).

Yup, as explained above that's considered as our "accepted legacy". It either should be cleaned up, or replaced by new code that is better/cleaner. I believe it's a good example why I aim to discuss and design new stuff on an issue first, with the goal of preventing to invest effort on repeating the past's bad patterns.

Sorry about still having it around, and causing to send you down on a seemingly less-than-ideal path for adding new `needs` tests :(

On the bright side, we at least gained more understanding of the problem domain. Let's find out the new best way together to test `needs`!

**Comment by tabulon** [comment]
> On the bright side, we at least gained more understanding of the problem domain. Let's find out the new best way together to test `needs`!

OK, so how do you want to go about this? 

Based on our discussion, here's a suggested outline for a new PR (after rebasing and seeing what PerlCritic has to say ) :

- Rewrite the test script, name it `t/needs2.t` (functionally covering both the current `t/needs.t` and `t/issue/1508.t`, and then some)
- Introduce the same code changes as in this PR (the second solution only)
- Changes, ...

For the rewrite of test script, I guess the below is what we've got:
 
- Get rid of POD & do away with TODO marks
- Factor out `package T` into a separate module, e.g. `t/lib/TaskTrack.pm` or something (any suggestions for the name?)
- See what can be done about the unpacking logic `.. %{; pop } ...`
- Improve readibility depending on what PerlCritic has to say...

On the other hand, I would be quite reluctant to move `Rex::Module` and `Nested::Module` packages into separate modules, because they are part of what is being actually tested  --even if that means silencing PerlCritic 

Any other suggestons?
